### PR TITLE
feat!: adding MinimumStake parameter

### LINKED ISSUE #501: Adding MinimumStake to the consensus parameters
## Describe the problem you'd like to have solved

Adding `MinimumStake` to the consensus parameters prevents creating dummy validators, specially in the Mainnet.
Spammers might create thousands of validators with small stakes. This has no effect on the consensus mechanism or the sortition algorithm.

### Describe the ideal solution

Adding a new field in the [consensus parameters](https://github.com/pactus-project/pactus/blob/main/types/param/param.go) for `MinimumStake`. The default value can set to 1 PAC coin.

### PR DESCRIPTION
**Pull request by kehiy**
## Description

This PR adds the `MinimumStake` parameter to the [consensus parameters](https://github.com/pactus-project/pactus/blob/236b5cac9e124a3baea8162478ae01c958e2f607/types/param/param.go#L5) and the default value is set to 1 PAC coin (10^9).
Consequently, all bond transaction will reject if their value is less than the `MinimumStake`.

## Related issue(s)

- Fixes #501

BREAKING CHANGE

### CODE REVIEW

**Comment by Ja7ad** [comment]
> ## Description
> I added `MinimumStake` amount as 1 PAC coin.
> there is just one issue with the test and it is:
> the random amount and fee always going to be less than `MinimumStake` we should change the random fee and amount or maybe the test.
> 
> ## Related issue(s)
> - Fixes #501

Please send your PR on **mainnet** branch.

**Comment by kehiy** [comment]
> > ## Description
> > I added `MinimumStake` amount as 1 PAC coin.
> > there is just one issue with the test and it is:
> > the random amount and fee always going to be less than `MinimumStake` we should change the random fee and amount or maybe the test.
> > 
> > ## Related issue(s)
> > - Fixes #501
> 
> Please send your PR on **mainnet** branch.

@kehiy sure!
I will send this PR and the GC pr to mainnet branch, but I should notice that this PR need a small change in merge time....

File: types/param/param.go [hunk 1]
```diff
@@ -14,6 +14,7 @@ type Params struct {
 	FeeFraction               float64 `cbor:"9,keyasint"`
 	MinimumFee                int64   `cbor:"10,keyasint"`
 	MaximumFee                int64   `cbor:"11,keyasint"`
+	MinimumStake              int64   `cobr:"12,keyasint"`
 	MaximumStake              int64   `cbor:"12,keyasint"`
```
Comment (b00f) [inline types/param/param.go:18]:
```suggestion
	MaximumStake              int64   `cbor:"13,keyasint"`
```

**Comment by kehiy** [comment]
@b00f I commited yout sugget, but don't forget to solve test issue in mergeing time.

**Comment by b00f** [comment]
@kehiy 

I think we can fix test with small modification for `randomAmountAndFee`, like:

```go
func (td *testData) randomAmountAndFee(min int64, max int64) (int64, int64) {
	amt := td.RandInt64NonZero(max)
	for amt < min {
		amt = td.RandInt64NonZero(max)
	}

	fee := int64(float64(amt) * td.sandbox.Params().FeeFraction)
	return amt, fee
}
```

**Comment by kehiy** [comment]
@b00f 
I test it, but that doesn't work!

**Comment by kehiy** [comment]
@b00f some other tests going to be filled, can you review them?

File: execution/executor/transfer_test.go [hunk 2]
```diff
@@ -57,7 +61,7 @@ func TestExecuteTransferTx(t *testing.T) {
 	senderAddr, senderAcc := td.sandbox.TestStore.RandomTestAcc()
 	senderBalance := senderAcc.Balance()
 	receiverAddr := td.RandomAddress()
-	amt, fee := td.randomAmountAndFee(senderBalance)
+	amt, fee := td.randomAmountAndFee(td.sandbox.TestParams.MinimumFee, senderBalance)
```
Comment (b00f) [inline execution/executor/transfer_test.go:64]: Only set minimum for the Bond transactions. for other transaction keep it zero.

**Comment by b00f** [comment]
@kehiy Please use `make test` to check it locally. I can see some issues in tests.
Also, for non-bond transactions, set minimum to zero.

Thanks buddy

**Comment by kehiy** [comment]
@b00f I fixed it, the last issue is `pool_test.go`

File: execution/executor/transfer_test.go [hunk 3]
```diff
@@ -44,10 +44,14 @@ func (td *testData) checkTotalCoin(t *testing.T, fee int64) {
...
-	amt := td.RandInt64(max / 2)
+func (td *testData) randomAmountAndFee(min int64, max int64) (int64, int64) {
+	amt := td.RandInt64NonZero(max)
+	for amt < min {
+		amt = td.RandInt64NonZero(max)
+	}
+
 	fee := int64(float64(amt) * td.sandbox.Params().FeeFraction)
-	return amt, fee
+	return fee, amt
```
Comment (b00f) [inline execution/executor/transfer_test.go:54]: why??

File: execution/executor/bond_test.go [hunk 4]
```diff
@@ -80,6 +79,14 @@ func TestExecuteBondTx(t *testing.T) {
 		assert.Equal(t, errors.Code(err), errors.ErrInvalidPublicKey)
 	})
 
+	t.Run("Should fail, amount less than MinimumStake", func(t *testing.T) {
+		trx := tx.NewBondTx(td.stamp500000, senderAcc.Sequence()+1, senderAddr,
```
Comment (b00f) [inline execution/executor/bond_test.go:83]: I think we changed `stamp500000` to `randStamp`.

@kehiy run `make unit_test` locally.

File: execution/executor/transfer_test.go (hunk 3 above)
Comment (kehiy) [reply execution/executor/transfer_test.go:54]: that was return them reverse I think
but I can change order of them when call it

File: execution/executor/bond_test.go (hunk 4 above)
Comment (kehiy) [reply execution/executor/bond_test.go:83]: the errors is from state test

File: execution/executor/bond_test.go [hunk 5]
```diff
@@ -116,7 +123,7 @@ func TestBondInsideCommittee(t *testing.T) {
 	exe2 := NewBondExecutor(false)
 	senderAddr, senderAcc := td.sandbox.TestStore.RandomTestAcc()
 	senderBalance := senderAcc.Balance()
-	fee, amt := td.randomAmountAndFee(senderBalance)
+	amt, fee := td.randomAmountAndFee(0, senderBalance)
```
Comment (b00f) [inline execution/executor/bond_test.go:126]: Thanks for fixing this mistake.

File: execution/executor/transfer.go [hunk 6]
```diff
@@ -54,6 +54,9 @@ func (e *TransferExecutor) Execute(trx *tx.Tx, sb sandbox.Sandbox) error {
 		return errors.Errorf(errors.ErrInvalidSequence,
 			"expected: %v, got: %v", senderAcc.Sequence()+1, trx.Sequence())
 	}
+	if senderAcc.Balance() < pld.Amount+trx.Fee() {
```
Comment (b00f) [inline execution/executor/transfer.go:57]: There is same logic several line above, Remove that.
The point is, first check sequence then balance.

File: genesis/genesis_test.go [hunk 7]
```diff
@@ -57,8 +58,8 @@ func TestGenesisTestNet(t *testing.T) {
 	genTime, _ := time.Parse("2006-01-02", "2023-05-08")
 	assert.Equal(t, g.GenesisTime(), genTime)
 	assert.Equal(t, g.Params().BondInterval, uint32(120))
-
-	expected, _ := hash.FromString("d5b41d8c2d45a951329512915ae7b2f4a445aeda380a1d40c50b5a5743dfb3fa")
+	fmt.Println(g.Hash())
```
Comment (b00f) [inline genesis/genesis_test.go:61]: Remove this line, please

**Review by b00f — CHANGES_REQUESTED** [review]

File: genesis/genesis_test.go (hunk 7 above)
Comment (kehiy) [reply genesis/genesis_test.go:61]: fixed

File: execution/executor/transfer.go (hunk 6 above)
Comment (kehiy) [reply execution/executor/transfer.go:57]: fixed

**Review by themantre — APPROVED** [review]

**Comment by kehiy** [comment]
@b00f checks are passed!

**Pull request merged**